posts/promql-guide.md

```diff
@@ -0,0 +1,159 @@
+
+![](./images/promql-guide/irate.png)
+
+`increase(http_requests_total[1h])` - # of http requests in the last hour. This is equal to the `rate` * # of seconds.
+
+![](./images/promql-guide/increase.png)
+
+These are a small fraction of the functions, just what we found most popular. You can find the rest [here](https://prometheus.io/docs/prometheus/latest/querying/functions/).
+
+### For Instant Vectors
+
+You'll notice that `rate(http_requests_total[5m])` above provides a ton of data. You can filter that data using your labels, but you can also look at your system as a whole using `sum` (or do both).
+
+`sum(rate(http_requests_total[5m]))`
+
+You can also use `min`, `max`, `avg`, `count`, and `quintile` similarly.
+
+![](./images/promql-guide/sum-rate.png)
+
+This tells you how many total HTTP requests there are, but some are obviously more indicative than others.
+
+`sum by(status_code) (rate(http_requests_total[5m]))`
+
+You can also use `without` rather than `by` to sum on everything not passed as a parameter to without.
+
+![](./images/promql-guide/sum-by-rate.png)
+
+Now, you can see the difference between each status code.
+
+### Offset
+
+You can use `offset` to change the time for Instant and Range Vectors. This can be helpful when comparing current to past usage when determining to trigger an alert.
+
+`sum(rate(http_requests_total[5m] offset 5m))`
+
+Remember to put `offset` directly after the selector.
+
+## Operators
+
+Operators can be used between scalars, vectors, or a mix of the two. Operations between vectors expect to be able to find matching elements for each element on the other side (also known as one-to-one matching), unless otherwise specified.
+
+There are Arithmetic (+, -, \*, /, %, ^), Comparison (==, !=, >, <, >=, <=) and Logical (and, or, unless) operators.
+
+### Vector Matching
+
+#### One-to-One
+
+Vectors are equal i.f.f. the labels are equal.
+
+API 5xxs are 10% of HTTP Requests
+`rate(http_requests_total{status_code=~"5.*"}[5m]) > .1 * rate(http_requests_total[5m])`
+
+![](./images/promql-guide/api5xx.png)
+
+We're looking to graph whenever more than 10% of an instance's HTTP requests are errors. Before comparing rates, PromQL first checks to make sure that the vector's labels are equal.
+
+You can use `on` to compare using certain labels or `ignoring` to compare on all labels except.
+
+#### Many-to-One
+
+It's possible to use comparison and arithmetic operations where an element on one side can be matched with many elements on the other side. _You must explicitly tell Prometheus what to do with the extra dimensions._
+
+You can use `group_left` if the left side has a higher cardinality, else use `group_right`.
+
+## Examples
+
+_Disclaimer_: We've hidden some of the information in the pictures using the `Legend Format` for privacy reasons.
+
+_CPU Usage by Instance_
+
+`100 * (1 - avg by(instance)(irate(node_cpu{mode='idle'}[5m])))`
+
+This provides the average CPU Usage per instance for a 5 minute window.
```

Comment:
Average CPU Usage per instance for a 5 minute window.

Reply:
This looks the same as what I wrote, unless you think I should capitalize the first letter in 'average'

Reply:
Right, I'm saying remove `This provides` since an intro is not used for the other examples in the descriptions.